core/src/main/java/org/apache/iceberg/CatalogUtil.java

```diff
@@ -151,20 +175,34 @@ private static void deleteFiles(FileIO io, Set<ManifestFile> allManifests) {
         .run(
             manifest -> {
               try (ManifestReader<?> reader = ManifestFiles.open(manifest, io)) {
+                List<String> pathsToDelete = Lists.newArrayList();
                 for (ManifestEntry<?> entry : reader.entries()) {
                   // intern the file path because the weak key map uses identity (==) instead of
                   // equals
                   String path = entry.file().path().toString().intern();
                   Boolean alreadyDeleted = deletedFiles.putIfAbsent(path, true);
                   if (alreadyDeleted == null || !alreadyDeleted) {
                     try {
-                      io.deleteFile(path);
+                      if (io instanceof SupportsBulkOperations) {
```

Comment:
What about collecting files into a list in all cases and then checking what type of `FileIO` we got? Hopefully, it should be a little bit more readable and will avoid the instanceof call for each manifest entry, which is expensive.

```
if (io instanceof SupportsBulkOperations) {
  SupportsBulkOperations bulkIO = (SupportsBulkOperations) io;
  bulkDelete(bulkIO, pathsToDelete, "files");
} else {
  for (String path : pathsToDelete) {
    delete(io, path);
  }
}
```

Reply:
Sure, this looks more readable to me, updated!